### PR TITLE
integrate reporting API documentation

### PR DESCRIPTION
**Pull request by jmarien**
This is a work in progress, an attempt at including the OpenAPI json spec of Reporting API in our Jekyll codebase for docs.soda.io. Note the extra attribute `apidoc: true` in  `api-docs\reporting-api.md`, if that is set the JS and CSS for Elements web component (see https://meta.stoplight.io/docs/elements/ZG9jOjEyMDU2Njg0-web-component) is loaded in `_includes\head.html`. 

What remains to be discussed or investigated:

- [ ] there are 2 types of layout the generated docs, which one is the most useful or closest to what we want
- [ ] for the one with the sidebar, can we collapse the left sidebar by default so we have the best of both worlds (API navigation + the ability to navigate the rest of the docs) perhaps? If that would not work: can we in worst case fall back to the one with the generated API sidebar and remove the left navigation completely, just for pages that have this kind of API docs embedded

Stacked layout:
image

Sidebar layout:
image

In order to switch from one to the other, just change the `layout attribute` in `api-docs\reporting-api.md`:

```html
<elements-api
    apiDescriptionUrl="/api-docs/openapi.json"
    router="hash"
    hideMocking="false"
    layout="stacked"
/>
```

You can checkout the branch locally and spin up the site to see what it will look like. 

If the above is decided, we still need to fix a few more things, like:

- [ ] not loading the CSS and JS (see `_includes\head.html` from a third party location
- [ ] fix CORS header so the tryout blocks work (should be done in Reporting API)
- [ ] decide on updating the JSON file through CI (can be done afterwards, apart from this PR) to keep it in sync

### CODE REVIEW

**Comment by janet-can** [comment]
Thanks for setting this up to preview, @jmarien. I tried both and clicked around each to get a feel for the layouts and I am conflicted. Overall, I prefer sidebar, with one reservation: the double-sidenav.
(It looks terrific, BTW!)

Stacked PROS:
- better on mobile or tablet? (I haven't seen it, obviously, but it might be a more digestible layout on a device.)
- only one sidenav
- the wider layout means less text wrapping in body

Stacked CONS:
- there doesn't seem to be a menu for the Schemas the same way there is in the sidebar layout, though the schemas are embedded it appears in each endpoint
- it's slightly less intuitive to navigate relative to the sidebar

Sidebar PROS:
- familiar layout, intuitive to navigate
- can see all the schemas in list format in API sidebar nav

Sidebar CONS:
- two sidenavs is a deal-breaker for me. If we were to choose this option, we must then also find a way to collapse the docs sidenav
- perhaps not an ideal layout for mobile
- it is pretty wordwrap heavy, but it's not the end of the world.

**Comment by bastienboutonnet** [comment]
@dirkgroenen you have a solution in mind for the display of the columned view without the nav bar of docs.soda.io right?

**Comment by dirkgroenen** [comment]
> @dirkgroenen you have a solution in mind for the display of the columned view without the nav bar of docs.soda.io right?

Yes, will add it on Monday at its latest - don't think I can find the time for it today.

**Comment by dirkgroenen** [comment]
Hacked my way through the templates to try to give you what you asked for @bastienboutonnet. I think the result is the best we can get out of it without going too crazy on the hacking. Wydt? 

image
image

**Comment by bastienboutonnet** [comment]
@dirkgroenen I think some words are missing in your previous comment and I'm not sure what to answer.

**Comment by dirkgroenen** [comment]
Updated my previous comment with a GIF to demonstrate how it now looks @bastienboutonnet.

BTW also be aware that the Stoplight docs thingy has no responsive design included, so this will only work on desktop. [It's on their roadmap](https://roadmap.stoplight.io/c/162-responsive-layout-for-table-mobile).

**Comment by bastienboutonnet** [comment]
@dirkgroenen this looks pretty good to me!

Let's see what @janet-can thinks and I guess we could merge it.

However, @jmarien do we maybe want to make sure we have a CI solution in place before emerging?

**Comment by jmarien** [comment]
@bastienboutonnet I would first make sure we can deploy this and then hook up CI as well.

**Comment by bastienboutonnet** [comment]
@jmarien makes sense, I'll update the open api spec on this PR then to allow for CORS and improved documentation context to be deployed as well

**Comment by jmarien** [comment]
Yes, up for review by @janet-can

**Comment by janet-can** [comment]
Hi all, 
I've reviewed the contents and done an editing run-thru for grammar and style. Looks great! 
 
A couple of adjustments to make before publishing:

ui-notes

- move the breadcrumbs to the right to align with page title (as with other docs pages)
- Change "Back to docs" to "Back to Help Center"  (I tried to do this myself but broke sizing of the logo so I leave it to you!)
- Why is the label FastAPI applied to the nav and to the overview page?  Could we not use Reporting API..?

**Review by janet-can — APPROVED** [review]
All good! Just needs three adjustments-- see my comment with screenshot in the main Conversation.

**Comment by jmarien** [comment]
@dirkgroenen can you take care of the first 2 feedback points please and if so, do I keep this PR open so you can just push your changes here? @bastienboutonnet the FastAPI title (3rd remark) is coming straight from the openapi.json, I suspect?

**Comment by bastienboutonnet** [comment]
Yes @jmarien Title and description is indeed coming from the json something funky must be happening when I generate it. I'll have a look and correct it.

**Comment by bastienboutonnet** [comment]
@janet-can in the future changes to the openapi.json need to happen in the `sodadata/reporting` repo because that's where the `.json` is generated from. 

Since you made a few changes here, I don't want to throw away your work so I'll move them manually to the other repo but worth knowing for the future. This will of course be clearer when we set up less copy-paste workflow for deployment (i.e. the json will not be hosted here).

I can show you where to make amends to the docstrings and field descriptions on the reporting repo if you want for future reference.

File: api-docs/openapi.json [hunk 1]
```diff
@@ -6,8 +6,8 @@
     "/v0/ping": {
       "get": {
         "tags": ["Status"],
-        "summary": "Soda Can Is Open",
-        "description": "Ping endpoint to test wether the Reporting API is live.\nReturns the sound of a Soda can opening along with a 200.",
+        "summary": "Soda can is open",
```
Comment (bastienboutonnet) [inline api-docs/openapi.json:9]: This correction cannot be enforced as it is not complient with the Open API spec. This is automatically generated. Is this a hard requirement?

File: api-docs/openapi.json [hunk 2]
```diff
@@ -19,9 +19,9 @@
     },
     "/v0/adoption/sign_ups": {
       "post": {
-        "tags": ["Tool Adoption"],
-        "summary": "Sign Ups",
-        "description": "This endpoint returns sign up events for the authenticated user's organization.\n\nThis data is currently updated once a day, and does not include the current date (UTC) in order\nto avoid incomplete reporting.\n\n**NOTE:** Users who signed up via SSO are currently not tracked by our\nevent tracking solution. If you do not see any sign up this could be the reason why.\n**We are working on a solution.**\n\nUsage\n-----\nYou can use this endpoint to build reports on Soda's adoption across your organization.\n\nReturns\n-------\nAn object of type `SignUpsResult` contatining an array of `SignUpData`.\n\nYou can find details about the response objects in the \"Schemas\" section or by\nunfurling the response details.",
+        "tags": ["Tool adoption"],
```
Comment (bastienboutonnet) [inline api-docs/openapi.json:22]: these end up as titles, shouldn't we use title case?

**Comment by dirkgroenen** [comment]
@jmarien @janet-can I've made a few more changes in an attempt to line up everything as best as possible. See the below GIF for the result. 

I think we should leave it at this, since I'm hacking more and more things using ugly approaches which makes it very fragile for updates. 

@jmarien To make it less fragile I suggest we fix the dependencies coming from Spotlight, instead of the dynamic 'latest version' approach currently in use. Agree? 

screen-capture (15) webm

**Comment by janet-can** [comment]
> @janet-can in the future, can you submit changes to the API docs in to the openapi.json need to happen in the `sodadata/reporting` repo because that's where the docs are generated. I'm going to have to move all your changes manually there. But I really do appreciate your suggestions they all make a lot of sense.
> 
> I'll show you how to make changes to the API docs on our call today if you want so that you can make PRs directly in that repo.

Oh no! I'm sorry. Yes, of course I'll make all future changes directly in the openapi.json in the repo. Sorry for the extra work. I can move them over today, if you haven't already done so.

**Comment by janet-can** [comment]
In thinking about this overnight, we should write a different overview to replace the one coming from FastAPI. I can draft something, unless you would prefer to do so @bastienboutonnet?

**Comment by bastienboutonnet** [comment]
@janet-can it's my fault, I should have told you. No harm done. I figured it would be easier for me to move them in the meantime. I've merged my changes in this PR (https://github.com/sodadata/docs/pull/114/commits/f8588e27d8489e360ea2c33534f7dc80771cae06)

So it all is good, from my side it's good.

I also see that @dirkgroenen committed the last styling changes so I would say this PR is definitely ready for review and we'll wait for your final approval and merge @janet-can .

(Regarding reporting doc content changes, I'll take some time with you to tell you how you can propose changes in the future.)

**Comment by bastienboutonnet** [comment]
@janet-can regarding the "overview" I believe you mean the top description? If so, yes I totally agree and welcome your suggestions. I put something in there to get us started but I think you can do a lot better.

Since you have some ideas, I would say go ahead and draft it, and I can show you how to propose the changes in the reporting API codebase.

**Comment by janet-can** [comment]
Re: overview, yes that's great. I just opened the PR now and can see what you've written. You were way ahead of me!

**Comment by bastienboutonnet** [comment]
I'm sure it can be further improved though. The reason why you didn't see it is that the one that we tested with did not have the custom description.

**Comment by janet-can** [comment]
Update: All editing done, this PR is fine to merge if we want.  However, if we can wait until tomorrow, there are two items left:
- I made some changes to two files in the reporting repo but couldn't commit them because of a local error. I've sent those changes to @bastienboutonnet to kindly request his help in making the changes and merging. Not a blocker.
- Do we need release notes for this API? If so, I can quickly create something and commit here. Not a blocker.

**Comment by bastienboutonnet** [comment]
Let's wait a bit indeed:
- Your changes on the reporting API repo are merged @janet-can I'll make a new copy of the `.json` here.
- I'll be opening a PR to the repo with release notes for the API (it'll point to blog and step by step guide)
- I'll move the step by step to here.

I would advise the following merge order:
- If we publish the blog post today
    - merge this PR
    - merge the step by step guide PR
    - merge the release notes for reporting API PR
- if we don't publish today, same order but I'll have to modify the release notes to that it does not

**Pull request merged**